### ISSUE TITLE
[Contrib metrics] Add R2 score metric

### ISSUE THREAD

**Issue by vfdev-5** [pytorch/ignite#453]
Maybe adding [R2 metric](https://en.wikipedia.org/wiki/Coefficient_of_determination) to `ignite.contrib.metrics.regression` could make sense. Thoughts ?

**Comment by anmolsjoshi** [comment]
@vfdev-5 I think that's a good idea! R2 is probably the first regression metric people start with.

**Comment by anmolsjoshi** [comment]
@vfdev-5 do you think this should be an epoch metric? or does it make sense to calculate this batch by batch?

**Comment by vfdev-5** [comment]
@anmolsjoshi I would say batch per batch is feasible.

**Comment by anmolsjoshi** [comment]
@vfdev-5 sounds good, if you haven't started working on it, I can send in a PR tonight

**Comment by vfdev-5** [comment]
yes, go ahead, please :)

**Comment by silas2471** [comment]
Hey, is it also possible to use this for deep learning as loss function?
I couldnt find any documentation on an implementation yet. Is there an easy way to implement it for such a use case?

**Comment by vfdev-5** [comment]
@silas2471 in ignite R2 is implemented as a metric to compute in online fashion (accumulation batch by batch). This is not reusable as a loss function. However, R2 definition is not difficult to reimplement as a loss function for a single batch:
```python
import numpy as np
from sklearn.metrics import r2_score
import torch
import torch.nn as nn
from torch.nn import functional as F


class R2Loss(nn.Module):
    
    def forward(self, y_pred, y):
        var_y = torch.var(y, unbiased=False)
        return 1.0 - F.mse_loss(y_pred, y, reduction="mean") / var_y


size = 51
np_y_pred = np.random.rand(size,)
np_y = np.random.rand(size,)

criterion = R2Loss()
y_pred = torch.from_numpy(np_y_pred)
y = torch.from_numpy(np_y)

res = criterion(y_pred, y)
print(r2_score(np_y, np_y_pred), res.item())
```
